### PR TITLE
Add functions to the datatype uuid

### PR DESCRIPTION
**Pull request by hououou**
add simple functions and gen_random_uuid() function to the datatype uuid

### CODE REVIEW

File: src/common/random_engine.cpp [hunk 1]
```diff
@@ -0,0 +1,41 @@
+#include "common/random_engine.hpp"
+
+#include <random>
+
+#include "pcg_random.hpp"
+
+namespace kuzu {
+namespace common {
+struct RandomState {
```
Comment (acquamarin) [inline src/common/random_engine.cpp:9]: newline

File: src/include/common/random_engine.hpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
+#pragma once
```
Comment (acquamarin) [inline src/include/common/random_engine.hpp:1]: Rename this file to .h

File: src/include/common/random_engine.hpp [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+#include <mutex>
+#include <random>
+
+#include "common/assert.h"
+
+namespace kuzu {
+
```
Comment (acquamarin) [inline src/include/common/random_engine.hpp:8]: remove white space between namespace

File: src/include/common/random_engine.hpp [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+namespace main {
+class ClientContext;
+}
+
+namespace common {
+
+struct RandomState;
+
+struct RandomEngine {
+    explicit RandomEngine(int64_t seed = -1);
```
Comment (acquamarin) [inline src/include/common/random_engine.hpp:18]: I see we only create RandomEngine using zero parameter. Why do we keep this optional argument?

File: src/include/common/random_engine.hpp [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    ~RandomEngine();
+
+public:
+    //! Generate a random number between min and max
```
Comment (acquamarin) [inline src/include/common/random_engine.hpp:22]: Our comment style is `// xxxx`, so remove `!`

File: src/include/common/random_engine.hpp [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    double NextRandom(double min, double max);
+
+    //! Generate a random number between 0 and 1
+    double NextRandom();
```
Comment (acquamarin) [inline src/include/common/random_engine.hpp:26]: use camel case for function names

File: src/include/function/arithmetic/arithmetic_functions.h [hunk 7]
```diff
@@ -316,5 +317,11 @@ struct Pi {
     static inline void operation(double_t& result) { result = M_PI; }
 };
 
+struct Uuid {
```
Comment (acquamarin) [inline src/include/function/arithmetic/arithmetic_functions.h:320]: UUID is a short hand for `Universal Unique Identifier`, so either write all characters in upper case or in lower case.

File: src/include/function/arithmetic/arithmetic_functions.h [hunk 8]
```diff
@@ -316,5 +317,11 @@ struct Pi {
...
+    static inline void operation(common::uuid_t& result) {
```
Comment (acquamarin) [inline src/include/function/arithmetic/arithmetic_functions.h:321]: I think the operation name should be something like `generateRandomUUID`

File: src/common/random_engine.cpp [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+double RandomEngine::NextRandom() {
+    return std::ldexp(randomState->pcg(), -32);
+}
+uint32_t RandomEngine::NextRandomInteger() {
+    return randomState->pcg();
+}
+
+void RandomEngine::SetSeed(uint32_t seed) {
```
Comment (acquamarin) [inline src/common/random_engine.cpp:37]: Where do we use this function?

**Review by acquamarin — COMMENTED** [review]
General comment:
1. Use camel case for function/variable names
2. Remove functions/parameters that haven't been used.

File: src/include/function/arithmetic/arithmetic_functions.h [hunk 10]
```diff
@@ -5,6 +5,7 @@
 
 #include "common/exception/runtime.h"
 #include "common/types/int128_t.h"
+#include "common/types/uuid.h"
```
Comment (andyfengHKU) [inline src/include/function/arithmetic/arithmetic_functions.h:8]: double check if we need this

File: src/include/function/pointer_function_executor.h [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+#include "common/vector/value_vector.h"
+
+namespace kuzu {
+namespace main {
+class ClientContext;
+}
+
+namespace function {
+
+struct PointerFunctionExecutor {
```
Comment (andyfengHKU) [inline src/include/function/pointer_function_executor.h:12]: This executor doesn't take in any parameter. So I assume ConstFunctionExecutor is a better naming.

File: src/include/function/uuid/functions/gen_random_uuid.h [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+struct UUID;
+struct uuid_t;
+} // namespace common
+
+namespace function {
+
+struct GenRandomUUID {
+    static inline void operation(common::ValueVector& result, void* dataPtr) {
+        KU_ASSERT(result.state->isFlat());
+        auto resultValues = (common::uuid_t*)result.getData();
```
Comment (andyfengHKU) [inline src/include/function/uuid/functions/gen_random_uuid.h:21]: Double check with @acquamarin, if UUID is use STRING layout, we should add through `StringVector::addString` interface to handle overflow properly.

File: src/include/main/client_context.h [hunk 13]
```diff
@@ -73,6 +78,8 @@ class ClientContext {
 
     void setExtensionOption(std::string name, common::Value value);
 
+    inline common::RandomEngine& getRandomEngine() { return *randomEngine; }
```
Comment (andyfengHKU) [inline src/include/main/client_context.h:81]: I think we just return pointer. If you return a reference and `RandomEngine` implements a default copy constructor, someone might trigger copy unexpected through `auto engine = context.getRandomEngine()`

File: src/include/processor/plan_mapper.h [hunk 14]
```diff
@@ -34,59 +38,92 @@ class PlanMapper {
         : storageManager{storageManager}, memoryManager{memoryManager},
           expressionMapper{}, catalog{catalog}, physicalOperatorID{0} {}
 
-    std::unique_ptr<PhysicalPlan> mapLogicalPlanToPhysical(
-        planner::LogicalPlan* logicalPlan, const binder::expression_vector& expressionsToCollect);
+    std::unique_ptr<PhysicalPlan> mapLogicalPlanToPhysical(planner::LogicalPlan* logicalPlan,
```
Comment (andyfengHKU) [inline src/include/processor/plan_mapper.h:41]: Can we put clientContext as a member of PlanMapper?

File: src/common/random_engine.cpp [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+
+RandomEngine::RandomEngine() : randomState(std::make_unique<RandomState>()) {
+    randomState->pcg.seed(pcg_extras::seed_seq_from<std::random_device>());
+}
+
+RandomEngine::~RandomEngine() {}
+
+double RandomEngine::nextRandom(double min, double max) {
+    KU_ASSERT(max >= min);
+    return min + (nextRandom() * (max - min));
```
Comment (andyfengHKU) [inline src/common/random_engine.cpp:25]: double check the coverage

**Review by andyfengHKU — APPROVED** [review]

File: src/common/random_engine.cpp [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+
+#include <random>
+
+#include "common/assert.h"
+#include "pcg_random.hpp"
+
+namespace kuzu {
+namespace common {
+
+struct RandomState {
```
Comment (acquamarin) [inline src/common/random_engine.cpp:11]: Define the struct inside random_engine.h so you don't have to forward declare in .h file.
And we always put fields in front of the the methods in a struct.

File: src/include/common/random_engine.h [hunk 17]
```diff
@@ -0,0 +1,33 @@
...
+    double nextRandom(double min, double max);
+
+    // Generate a random number between 0 and 1
+    double nextRandom();
+    uint32_t nextRandomInteger();
+
+    std::mutex lock;
+
+private:
+    std::unique_ptr<RandomState> randomState;
```
Comment (acquamarin) [inline src/include/common/random_engine.h:30]: Guess you probably don't have to use unique_ptr there?

**Review by acquamarin — APPROVED** [review]

**Pull request merged**